### PR TITLE
test: Search PHPUnit tests

### PR DESCRIPTION
**Pull request by Kallyan01**
## What
Added Search Module PHPUnit test cases.

## Why

Improves development workflow by enabling faster and more reliable validation of REST functionality, reducing repetitive manual testing efforts.

### Related Issue(s):

- https://github.com/rtCamp/OnePress/issues/71
### AI Disclosure

Hybrid approach = Written by Me + Copilot (Claude Opus 4.6, GPT 5.3) + Audited by me

## Testing Instructions

```bash
npm run wp-env:test -- start --xdebug=coverage    

npm run test:php -- tests/php/Unit/Modules/Search
```

## Checklist

- [ ] I have read the [Contribution Guidelines](https://github.com/rtCamp/onesearch/blob/main/docs/CONTRIBUTING.md).
- [ ] I have read the [Development Guidelines](https://github.com/rtCamp/onesearch/blob/main/docs/DEVELOPMENT.md).
- [ ] I have added necessary tests to cover my changes.
- [ ] I have updated the project documentation as needed.
- [ ] My code has detailed inline documentation.
- [ ] My code is tested to the best of my abilities.
- [ ] My code passes all lints, tests, and checks.

<a href="https://playground.wordpress.net?blueprint-url=data:application/json,%7B%22%24schema%22%3A%22https%3A%2F%2Fplayground.wordpress.net%2Fblueprint-schema.json%22%2C%22meta%22%3A%7B%22title%22%3A%22OneSearch%20Demo%22%2C%22description%22%3A%22Cross-site%2C%20brand-aware%20search%20across%20a%20multi-brand%20WordPress%20network%2C%20powered%20by%20Algolia.%22%2C%22author%22%3A%22rtCamp%22%2C%22categories%22%3A%5B%22plugin%22%2C%22search%22%2C%22multisite%22%2C%22algolia%22%2C%22enterprise%22%5D%7D%2C%22landingPage%22%3A%22%2Fwp-admin%2Fplugins.php%22%2C%22preferredVersions%22%3A%7B%22php%22%3A%228.4%22%2C%22wp%22%3A%22latest%22%7D%2C%22features%22%3A%7B%22networking%22%3Atrue%7D%2C%22steps%22%3A%5B%7B%22step%22%3A%22login%22%2C%22username%22%3A%22admin%22%7D%2C%7B%22step%22%3A%22installPlugin%22%2C%22pluginData%22%3A%7B%22resource%22%3A%22url%22%2C%22url%22%3A%22https%3A%2F%2Fgithub.com%2FrtCamp%2Fonesearch%2Freleases%2Fdownload%2Fci-artifacts%2Fpr-213-31ac94da8316df831b3516328b9ecee68afdc98a.zip%22%7D%2C%22options%22%3A%7B%22activate%22%3Atrue%7D%7D%5D%7D" target="_blank" rel="noopener noreferrer">
  Open WordPress Playground Preview
</a>

### CODE REVIEW

File: tests/php/Unit/Modules/Search/SettingsTest.php [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+
+		$sanitized = $sanitize(
+			[
+				'app_id'    => ' <b>app-id</b> ',
+				'write_key' => "\n<script>alert(1)</script>write-key\t",
+			]
+		);
+
+		$this->assertSame( 'app-id', $sanitized['app_id'] );
+		$this->assertSame( 'write-key', $sanitized['write_key'] );
```
[the remark this answers is not in the document]
Comment (Kallyan01) [reply tests/php/Unit/Modules/Search/SettingsTest.php:75]: Disagree

File: tests/php/Unit/Modules/Search/WatcherTest.php [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+			]
+		);
+
+		$post    = self::factory()->post->create_and_get( [ 'post_status' => 'publish' ] );
+		$watcher = new Watcher();
+
+		// Transitioning to 'trash' — should not reindex (only delete).
+		$watcher->on_post_transition( 'trash', 'publish', $post );
+
+		$this->assertTrue( true );
```
[the remark this answers is not in the document]
Comment (Kallyan01) [reply tests/php/Unit/Modules/Search/WatcherTest.php]: The suggested code changes won't work here as Algolia SDK doesn't use WordPress's HTTP API (wp_remote_*), it has its own HTTP layer. The pre_http_request filter only intercepts wp_remote_* calls, so it never fires for Algolia's delete_by(). The fix is to use AlgoliaSDK::setHttpClient()

File: tests/php/Unit/Modules/Search/IndexTest.php [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	public function test_index_all_posts_returns_error_when_delete_fails(): void {
+		update_option( Settings::OPTION_SITE_TYPE, Settings::SITE_TYPE_GOVERNING );
+		delete_option( Search_Settings::OPTION_GOVERNING_ALGOLIA_CREDENTIALS );
+
+		$result = ( new Index() )->index_all_posts( [ 'post' ] );
+
+		$this->assertWPError( $result );
+		$this->assertSame( 'algolia_credentials_missing', $result->get_error_code() );
+	}
+}
```
Comment (sabbir1991) [inline tests/php/Unit/Modules/Search/IndexTest.php:377]: All the tests you did the error (missing credentials) part. Do you think we need at least one successful assertion for each function?

File: tests/php/Unit/Modules/Search/SearchTest.php [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+		$remote_post                        = new \WP_Post( new \stdClass() );
+		$remote_post->onesearch_original_id = 17;
+		$remote_post->guid                  = 'https://remote.example.com/posts/17/';
+
+		$wp_query->posts = [ $remote_post ];
+
+		$search = new Search();
+		$result = $search->get_post_type_permalink( 'https://example.com/local/', -18 );
+
+		$this->assertSame( 'https://remote.example.com/posts/17/', $result );
```
Comment (sabbir1991) [inline tests/php/Unit/Modules/Search/SearchTest.php:278]: I think we need to set the $remote_post_id to -18. or $onesearch_original_id should be 18. If in production someone use abs(-18), then maybe the test will fail. Can you please verify again and share your feedback

File: tests/php/Unit/Modules/Search/SearchTest.php [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+	 * Delegates to get_term_link with 'post_tag' taxonomy.
+	 */
+	public function test_get_tag_link_returns_default_when_search_disabled(): void {
+		delete_option( Search_Settings::OPTION_GOVERNING_SEARCH_SETTINGS );
+
+		$search = new Search();
+		$result = $search->get_tag_link( 'https://example.com/tag/php/', 3 );
+
+		$this->assertSame( 'https://example.com/tag/php/', $result );
+	}
```
Comment (sabbir1991) [inline tests/php/Unit/Modules/Search/SearchTest.php:524]: Both tests are covered for search_disable for specific category and tags, do you think we need to test also enabled one. or you are coverting that test using this function https://github.com/rtCamp/onesearch/pull/213/changes#diff-e745a58f87fabf2cd35e27bf1944cf8ebe734fc19e7845875d16aefeee7ce1b4R367

File: tests/php/Unit/Modules/Search/SearchTest.php [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+	 */
+	public function test_get_post_terms_returns_original_when_search_disabled(): void {
+		delete_option( Search_Settings::OPTION_GOVERNING_SEARCH_SETTINGS );
+
+		$search   = new Search();
+		$original = [ new \WP_Term( new \stdClass() ) ];
+
+		$result = $search->get_post_terms( $original, 1, 'category' );
+		$this->assertSame( $original, $result );
+	}
```
Comment (sabbir1991) [inline tests/php/Unit/Modules/Search/SearchTest.php:564]: Same here only disable, no enable search test

**Review by sabbir1991 — COMMENTED** [review]
Please review the coverage HTML again and check the functions are fully covered which test you have written

File: tests/php/Unit/Modules/Search/IndexTest.php (hunk 3 above)
Comment (Kallyan01) [reply tests/php/Unit/Modules/Search/IndexTest.php:377]: I avoided this initially since it relied on Algolia runtime behaviour and could be flaky without a clean mock seam. With our Algoila SDK HTTP mock in WatcherTest.php, we can safely add this coverage

File: tests/php/Unit/Modules/Search/SearchTest.php (hunk 4 above)
Comment (Kallyan01) [reply tests/php/Unit/Modules/Search/SearchTest.php:278]: Placeholder remote post ID is computed as -1 -post_id = post_id, so post_id 17 becomes -18.

**Review by sabbir1991 — DISMISSED** [review]

File: tests/php/Unit/Modules/Search/IndexTest.php [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+#[CoversClass( \OneSearch\Modules\Search\Index::class )]
+final class IndexTest extends TestCase {
+	/**
+	 * {@inheritDoc}
+	 */
+	protected function tearDown(): void {
+		AlgoliaSDK::resetHttpClient();
+
+		delete_option( Settings::OPTION_SITE_TYPE );
+		delete_option( Search_Settings::OPTION_GOVERNING_ALGOLIA_CREDENTIALS );
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/IndexTest.php]: I don't think these are needed since everything happens in between `setUp()` and `tearDown()` 
```suggestion

```

File: tests/php/Unit/Modules/Search/IndexTest.php [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+		$result = ( new Index() )->delete_index();
+
+		$this->assertWPError( $result );
+		$this->assertSame( 'algolia_credentials_missing', $result->get_error_code() );
+	}
+
+	/**
+	 * Returns true for delete_index with valid credentials.
+	 */
+	public function test_delete_index_returns_true_with_valid_credentials(): void {
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/IndexTest.php:90]: Per the coverage report, I think we want to also call once with `Settings::SITE_TYPE_CONSUMER`

File: tests/php/Unit/Modules/Search/SearchTest.php [hunk 9]
```diff
@@ -0,0 +1,688 @@
...
+	/**
+	 * {@inheritDoc}
+	 */
+	protected function tearDown(): void {
+		// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited, SlevomatCodingStandard.Variables.UnusedVariable.UnusedVariable -- Resetting test state.
+		global $post, $wp_query, $wp_the_query;
+
+		$post         = null; // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+		$wp_query     = null; // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited, SlevomatCodingStandard.Variables.UnusedVariable.UnusedVariable
+		$wp_the_query = null; // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited, SlevomatCodingStandard.Variables.UnusedVariable.UnusedVariable
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/SearchTest.php]: I don't think these are needed either. Maybe the transient (test be removing the `delete_transient` and seeing if everything still passes.)

File: tests/php/Unit/Modules/Search/SearchTest.php [hunk 10]
```diff
@@ -0,0 +1,688 @@
...
+
+	/**
+	 * Ensures register_hooks adds block render filter.
+	 */
+	public function test_register_hooks_adds_render_block_filter(): void {
+		$search = new Search();
+		$search->register_hooks();
+
+		$this->assertNotFalse( has_filter( 'render_block', [ $search, 'filter_render_block' ] ) );
+	}
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/SearchTest.php]: I don't think any of these are necessary. We're testing outcomes, not implementation details, so it's enough to confirm that `$search->register_hooks();` doesn't throw an exception.


```suggestion

	/**
	 * Ensures class can be instantiated.
	 */
	public function test_class_instantiation(): void {
		$search = new Search();
		$search->register_hooks();

		// If we made it this far, we're good 🍻
		$this->assertTrue( true );
	}
```

File: tests/php/Unit/Modules/Search/SearchTest.php [hunk 11]
```diff
@@ -0,0 +1,688 @@
...
+		);
+
+		$recorded_paths = [];
+		$this->mock_algolia_http_client(
+			$recorded_paths,
+			static function ( string $path ): string {
+				if ( str_contains( $path, '/query' ) ) {
+					return wp_json_encode(
+						[
+							'hits'        => [
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/SearchTest.php:108]: Can we create one more test with 3 (out of order) hits? I don't care to test the specific implementation details inside `::compute_algolia_score()`, just confirm that it's running and ordering them without throwing any notices.

File: tests/php/Unit/Modules/Search/SettingsTest.php [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+
+		$this->assertArrayHasKey( Search_Settings::OPTION_GOVERNING_ALGOLIA_CREDENTIALS, $registered );
+		$this->assertArrayHasKey( Search_Settings::OPTION_GOVERNING_INDEXABLE_SITES, $registered );
+		$this->assertArrayHasKey( Search_Settings::OPTION_GOVERNING_SEARCH_SETTINGS, $registered );
+	}
+
+	/**
+	 * Ensures register_settings sanitizes algolia credentials payload.
+	 */
+	public function test_register_settings_sanitizes_algolia_credentials(): void {
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/SettingsTest.php:62]: No need to test the implementation details by calling out the callback directly. Instead just `update_option()`, and then `get_option( * )` to confirm our sanitization is working.

File: tests/php/Unit/Modules/Search/SettingsTest.php [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+		];
+		update_option( Search_Settings::OPTION_GOVERNING_SEARCH_SETTINGS, $value );
+
+		$this->assertSame( $value, Search_Settings::get_search_settings() );
+	}
+
+	/**
+	 * Does nothing when new value is not consumer.
+	 */
+	public function test_on_site_type_change_skips_non_consumer(): void {
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/SettingsTest.php:219]: We also need to test it when it does work though.

File: tests/php/Unit/Modules/Search/WatcherTest.php [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+use OneSearch\Tests\TestCase;
+use OneSearch\Utils;
+use OneSearch\Vendor\Algolia\AlgoliaSearch\Algolia as AlgoliaSDK;
+use PHPUnit\Framework\Attributes\CoversClass;
+
+/**
+ * Tests for the Watcher class.
+ */
+#[CoversClass( \OneSearch\Modules\Search\Watcher::class )]
+final class WatcherTest extends TestCase {
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/WatcherTest.php:25]: You tested all the edge cases, but is there a way to test the happy path with a post from the `factory()`?

**Review by justlevine — CHANGES_REQUESTED** [review]
Great first pass! 🙌 Especially a fan of https://github.com/rtCamp/onesearch/pull/213/commits/daab8dcab2101f9dd25130bbf7c5cd41cd3e183b before I even asked ;-) 

Just some minor cleanup and trying to cover the few areas the coverage report missed. We'll be refactoring soon though, so if something I asked for isn't easily testable, lmk and maybe we'll skip it.

File: tests/php/Unit/Modules/Search/WatcherTest.php (hunk 14 above)
Comment (Kallyan01) [reply tests/php/Unit/Modules/Search/WatcherTest.php:25]: Tried to mock it by using mock_algolia_http_client().

File: tests/php/Unit/Modules/Search/IndexTest.php [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+		$result = ( new Index() )->index_all_posts( [ 'post' ] );
+
+		$this->assertWPError( $result );
+		$this->assertSame( 'algolia_credentials_missing', $result->get_error_code() );
+	}
+
+	/**
+	 * Returns true when index_all_posts is called with no post types and valid credentials.
+	 */
+	public function test_index_all_posts_returns_true_with_valid_credentials_and_no_post_types(): void {
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/IndexTest.php:294]: I think we need one more successful `index_all_posts` with 101 posts, so it triggers `::generate_post_batches()`, but I'm on the fence.

 @Kallyan01 thoughts?

**Review by justlevine — COMMENTED** [review]
These look good! 

I'm wondering if it's worth it for you to do one more pass through the coverage report, and see if there's some more low-hanging you can catch here, e.g. from a brand site, w. author info, composite scores, > 100 posts, etc. Still outcome based, nothing that requires elaborate stubbing.

Alternatively, we can skip them now add the coverage to the areas that we refactor, the missing areas aren't particularly important for back-compat, just catching unintentional breaks... 

@Kallyan01 can you take a look at the report and let me know which - if any - you think are worth doing now vs later?

File: tests/php/Unit/Modules/Search/IndexTest.php (hunk 15 above)
Comment (Kallyan01) [reply tests/php/Unit/Modules/Search/IndexTest.php:294]: Added the case, I think it's a good testcase to add.

File: tests/php/TestCase.php [hunk 16]
```diff
@@ -81,5 +81,68 @@ protected function checkRequirements(): void { // phpcs:ignore Generic.CodeAnaly
...
+	 * Intercept Algolia SDK HTTP calls and collect request paths.
+	 *
+	 * @param array<int, string>              $recorded_paths Paths captured from outgoing SDK requests.
+	 * @param (callable(string): string)|null $body_for_path Optional callback to provide a response body for a given request path.
+	 * @param string|null                     $throw_on_path_segment Optional path segment that triggers a RuntimeException when matched.
+	 */
+	public function mock_algolia_http_client( array &$recorded_paths, ?callable $body_for_path = null, ?string $throw_on_path_segment = null ): void {
+		\OneSearch\Vendor\Algolia\AlgoliaSearch\Algolia::setHttpClient(
+			new class( $recorded_paths, $body_for_path, $throw_on_path_segment ) implements \OneSearch\Vendor\Algolia\AlgoliaSearch\Http\HttpClientInterface {
+				/** @var array<int, string> */
```
Comment (Utsav-Ladani) [inline tests/php/TestCase.php:94]: Is it possible to move this class definition into a separate file?

**Review by Utsav-Ladani — DISMISSED** [review]
CI checks are failing due to GH token issue else LGTM.

File: tests/php/TestCase.php (hunk 16 above)
Comment (Kallyan01) [reply tests/php/TestCase.php:94]: @Utsav-Ladani We can do that, but since the mock_algolia_http_client method is only useful in the test environment and provides no practical benefit in production, it would simply add another file to the codebase without delivering any meaningful value.

@justlevine any thoughts?

Comment (justlevine) [reply tests/php/TestCase.php:94]: @Kallyan01 I agree with you.

At least currently, I don't see enough here to warrant creating an explicit test fixture or whatnot, and even if/when we decide to colocate all the plugins in a monorepo, we're still not likely to reusing the underlying `TestCase`.

For now I think we keep this as-is, and reevaluate once other test methods emerge that need to be shared that would justify it.

Comment (justlevine) [reply tests/php/TestCase.php:94]: @Utsav-Ladani if you feel strongly about separating this out, please feel free to continue the discussion and we can create a follow-up PR.

Comment (Utsav-Ladani) [reply tests/php/TestCase.php:94]: > but since the mock_algolia_http_client method is only useful in the test environment and provides no practical benefit in production, it would simply add another file to the codebase without delivering any meaningful value.

Then ignore keep it as it is.

File: tests/php/Unit/Modules/Search/SettingsTest.php [hunk 17]
```diff
@@ -0,0 +1,413 @@
...
+		$this->assertNotFalse(
+			has_action( 'update_option_' . Search_Settings::OPTION_GOVERNING_ALGOLIA_CREDENTIALS, [ $settings, 'purge_cache_on_update' ] )
+		);
+		$this->assertNotFalse(
+			has_action( 'update_option_' . Search_Settings::OPTION_GOVERNING_INDEXABLE_SITES, [ $settings, 'purge_cache_on_update' ] )
+		);
+		$this->assertNotFalse(
+			has_action( 'update_option_' . Search_Settings::OPTION_GOVERNING_SEARCH_SETTINGS, [ $settings, 'purge_cache_on_update' ] )
+		);
+	}
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/SettingsTest.php:41]: Unless the action is only conditionally enqueued, we never need to test that `add_action()` or `add_filter()` work. Checking if specific functions are hooked is fragile and unnecessary.

```suggestion

	/**
	 * Ensures register_hooks adds expected actions.
	 */
	public function test_class_instantiation(): void {
		$settings = new Search_Settings();
		$settings->register_hooks();

		// If we made it here without erroring, we're good.
		$this->assertTrue( true );
	}
```

File: tests/php/Unit/Modules/Search/WatcherTest.php [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+
+	/**
+	 * Ensures register_hooks adds transition_post_status action.
+	 */
+	public function test_register_hooks_adds_transition_post_status_action(): void {
+		$watcher = new Watcher();
+		$watcher->register_hooks();
+
+		$this->assertNotFalse( has_action( 'transition_post_status', [ $watcher, 'on_post_transition' ] ) );
+	}
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/WatcherTest.php:43]:
```suggestion
	/**
	 * Ensures the class is initialized correctly
	 */
	public function test_class_instantiation(): void {
		$watcher = new Watcher();
		$watcher->register_hooks();

		// If we made it this far, we're good.
		$this->assertTrue( true );
	}
```

**Review by justlevine — CHANGES_REQUESTED** [review]
Last two open comments, and this should be good to merge.

> If needed, we can add additional tests to cover the logic implemented in the private methods as well, like compute_algolia_score, extract_algolia_highlights, build_post_from_record, split_content_into_chunks, etc. Let me know your thoughts.

I think we've got enough in place for now, especially as its not fully clear to me which of those implementation details will be necessary in the future. We can backfill them as we update Algolia to v4 and our wrappers accordingly. 🚀

File: tests/php/Unit/Modules/Search/PostRecordTest.php [hunk 19]
```diff
@@ -185,6 +184,59 @@ public function test_split_content_into_chunks_splits_oversized_content_into_mul
...
+	public function test_split_content_into_chunks_trims_content(): void {
+		$record = new Post_Record();
+		$method = new \ReflectionMethod( Post_Record::class, 'split_content_into_chunks' );
+
+		$chunks = $method->invoke( $record, "  \t\n  hello  \n\t  ", 100 );
+
+		$this->assertCount( 1, $chunks );
+		$this->assertSame( 'hello', $chunks[0] );
+	}
+
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/PostRecordTest.php]: Revert these please they're implementation details.

(Maybe we need to rename `tests/php/Unit` to `tests/php/Integration` to make it more clear 🤔).

The way to cover these is as I said in previous comments: use the factory to create enough new records, and then update one to have enough to be chunked, then test through the _public_ methods.

But like I said [before](https://github.com/rtCamp/onesearch/pull/213#pullrequestreview-4485962274), I'm fine coming back to these after the upgrade to Algolia v4.

File: tests/php/Unit/Modules/Search/SearchTest.php [hunk 20]
```diff
@@ -675,6 +675,272 @@ public function test_filter_render_block_rewrites_excerpt_for_remote_post(): voi
...
+			]
+		);
+
+		$this->assertInstanceOf( \WP_Post::class, $result );
+		$this->assertSame( -9, $result->ID );
+		$this->assertFalse( isset( $result->onesearch_remote_post_author_display_name ) );
+		$this->assertCount( 1, $result->onesearch_remote_taxonomies );
+		$this->assertSame( 'News', $result->onesearch_remote_taxonomies[0]['name'] );
+	}
+
```
Comment (justlevine) [inline tests/php/Unit/Modules/Search/SearchTest.php]: Revert, same as [above](https://github.com/rtCamp/onesearch/pull/213#discussion_r3414625340). 

Tip: If you cannot access it directly, it's usually not something that should be tested directly.

**Review by justlevine — CHANGES_REQUESTED** [review]
@Kallyan01 I'm confused about https://github.com/rtCamp/onesearch/pull/213/commits/ebe9604f0d01065cf1206361a889e402b30116c2 and the two open comments.

As noted [in the last review](https://github.com/rtCamp/onesearch/pull/213#pullrequestreview-4485962274) literally all you had to do was Apply Suggestions to the two open comments and instead it's now two rounds review.

What am I missing?

**Review by justlevine — APPROVED** [review]
🥳
